### PR TITLE
added the SAG stochastic gradient descent method

### PR DESCRIPTION
**Pull request by Ulfgard**
I added the SAG method for linear models, see

https://hal.inria.fr/hal-00860051/document

it has linear convergence rate on strongly convex functions, otherwise O(1/T). I could not find its convergence rate on non-smooth functions(if it converges at all).

The implementation works for arbitrary data and loss-functions, especially dense & sparse vectors, classification & regression and weighted & un-weighted.

### CODE REVIEW

File: Test/Algorithms/Trainers/LinearSAGTrainer.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+ * \brief       test case for the Linear SAG-Trainer
+ * 
+ * 
+ * 
+ *
+ * \author      O.Krause
+ * \date        2016
+ *
+ *
+ * \par Copyright 1995-2015 Shark Development Team
```
Comment (TGlas) [inline Test/Algorithms/Trainers/LinearSAGTrainer.cpp]: 2015 -> 2016
We should really have a yearly cron job somewhere to automate this.

File: include/shark/Algorithms/Trainers/LinearSAGTrainer.h [hunk 2]
```diff
@@ -0,0 +1,276 @@
+//===========================================================================
+/*!
+ *
+ *
+ * \brief       Generic Stochastic Average Gradient Descent training for linear models
+ *
+ *
+ *
+ *
+ * \author      T. Glasmachers
```
Comment (TGlas) [inline include/shark/Algorithms/Trainers/LinearSAGTrainer.h]: Is it me? copy&paste...? :)

**Comment by TGlas** [comment]
This is great, I really love the SAG algorithm!
There is only one detail, the loop in the trainer is terminated based on the number of epochs. Can't we replace that with an AbstractStoppingCriterion?

**Comment by Ulfgard** [comment]
Hi,

the number of epochs is one of two possible stopping criteria in this case: number of epochs or change of point. as the change in function value or in gradient is an incredible noisy (and biased) statistic they do not make much sense. Change of point does not tell much about change of function value in ill-conditioned problems, so epochs it is.

Also this is the same stopping criterion as for the KernelSGDTrainer and I tried to match interfaces as close as possible.

should I also add a KernelSAGTrainer? thats straight forward to add.

Do you know any reference telling about convergence in the non-smooth case? I would like to add a comment about this.

**Comment by TGlas** [comment]
> the number of epochs is one of two possible stopping criteria in this case: number of epochs or change of point. as the change in function value or in gradient is an incredible noisy (and biased) statistic they do not make much sense. Change of point does not tell much about change of function value in ill-conditioned problems, so epochs it is.
Okay, I see.

> Also this is the same stopping criterion as for the KernelSGDTrainer and I tried to match interfaces as close as possible.
Makes sense.

> should I also add a KernelSAGTrainer? thats straight forward to add.
I already thought about doing this myself :)
In my experience a "brute force" kernel trainer does not make too
much sence, but it is straightforward. If possible then please keep
the set of support vectors independent of the training data.
Then we can throw in the training data as a special case (standard
kernel setting), but we can also use a random projection / budget
approach.

> Do you know any reference telling about convergence in the non-smooth case? I would like to add a comment about this.
No. As far as I remember they simply don't say anything about this
case in the papers.

Look at this example problem (with optimum x^* = 0):
    min |x+1| + |x| + |x-1|
Let's assume that we have already converged to the area -1 <= x <= +1,
then the agerage gradient coincides with the gradient of |x|. So why
should it be better than plain SGD?

Practically speaking, with hinge loss it does not seem to converge
any faster than SGD, so for an SVM one should really use a squared
hinge or Huber-type loss.

**Comment by Ulfgard** [comment]
> I already thought about doing this myself :) In my experience a "brute force" kernel trainer does not make too much sence, but it is straightforward. If possible then please keep the set of support vectors independent of the training data. Then we can throw in the training data as a special case (standard kernel setting), but we can also use a random projection / budget approach.

I thought about a version that is consistent with the kernel-SGD trainer. i.e. take the trainer and just replace the SGD part by SAG. So whatever we add, we should also add it to that trainer (or remove/replace/amend it). This means that we have the 2-norm regularization of the alpha-values instead of norm in hilbert-space, etc. If you want an independent basis, i can add that too. 

for the case of the L2-hinge loss we can add a "nice" caching strategy: only cache entries which have nonzero gradD components (they do not violate the margin -> gradient 0 and are also likely to be correct later on)

**Comment by TGlas** [comment]
>> I already thought about doing this myself :) In my experience a "brute force" kernel trainer does not make too much sence, but it is straightforward. If possible then please keep the set of support vectors independent of the training data. Then we can throw in the training data as a special case (standard kernel setting), but we can also use a random projection / budget approach.
>
> I thought about a version that is consistent with the kernel-SGD trainer. i.e. take the trainer and just replace the SGD part by SAG. So whatever we add, we should also add it to that trainer (or remove/replace/amend it). This means that we have the 2-norm regularization of the alpha-values instead of norm in hilbert-space, etc. If you want an independent basis, i can add that too.

That would be great, I believe that this is a must for large-scale
learning.

> for the case of the L2-hinge loss we can add a "nice" caching strategy: only cache entries which have nonzero gradD components (they do not violate the margin -> gradient 0 and are also likely to be correct later on)

Yes. This is analogous to shrinking in the dual case, and you could
even do hard shrinking. Alternatively you can select those guys less
frequently. In my experience this can help, but it can also destroy
some of the stability of the method.
Exploiting the structure "only" at the level of the kernel cache
sounds like a good compromise.

Tobias
